toolbox/library/src/OSQP.cpp

```diff
@@ -0,0 +1,589 @@
+            pImpl->totalConstraintsUpperBounds.tail(pImpl->numberOfVariables) = pImpl->variableConstraintsUpperBounds;
+        }
+
+        if (pImpl->useLbA || pImpl->useUbA) {
+            pImpl->totalConstraintsMatrix.topRows(pImpl->numberOfProperConstraints) = pImpl->properConstraintMatrix;
+        }
+
+        if (pImpl->useLb || pImpl->useUb) {
+            pImpl->totalConstraintsMatrix.bottomRows(pImpl->numberOfVariables) = Eigen::MatrixXd::Identity(pImpl->numberOfVariables, pImpl->numberOfVariables);
+        }
+
+        if (!pImpl->sqSolver->isInitialized()) {
+            pImpl->totalConstraintsMatrixSparse = pImpl->totalConstraintsMatrix.sparseView();
+            pImpl->sqSolver->data()->setLinearConstraintsMatrix(pImpl->totalConstraintsMatrixSparse);
+            pImpl->sqSolver->data()->setBounds(pImpl->totalConstraintsLowerBounds, pImpl->totalConstraintsUpperBounds);
+        } else {
+            pImpl->totalConstraintsMatrixSparse = pImpl->totalConstraintsMatrix.sparseView();
+            pImpl->sqSolver->updateLinearConstraintsMatrix(pImpl->totalConstraintsMatrixSparse);
+            pImpl->sqSolver->updateBounds(pImpl->totalConstraintsLowerBounds, pImpl->totalConstraintsUpperBounds);
+        }
+    }
+    // OUTPUTS
+    // =======
+
+    OutputSignalPtr solutionSignal = blockInfo->getOutputPortSignal(OutputIndex::PrimalSolution);
+    if (!solutionSignal) {
+        bfError << "Output signal not valid.";
+        return false;
+    }
+    Eigen::Map<Eigen::VectorXd> solution_colMajor(
+        const_cast<double*>(solutionSignal->getBuffer<double>()),
+        blockInfo->getOutputPortWidth(OutputIndex::PrimalSolution));
+
+
+    OutputSignalPtr statusSignal = blockInfo->getOutputPortSignal(OutputIndex::Status);
+    if (!statusSignal) {
+        bfError << "Status signal not valid.";
+        return false;
+    }
+    Eigen::Map<Eigen::VectorXd> status_colMajor(
+        const_cast<double*>(statusSignal->getBuffer<double>()),
+        blockInfo->getInputPortWidth(OutputIndex::Status));
+
+    // Set inputs
+    Eigen::SparseMatrix<double> hessian_sparse = hessian_colMajor.sparseView();
+
+    if (!pImpl->sqSolver->isInitialized()) {
+        pImpl->sqSolver->data()->setHessianMatrix(hessian_sparse);
+        pImpl->sqSolver->data()->setGradient(gradient_colMajor);
+    } else {
+        pImpl->sqSolver->updateHessianMatrix(hessian_sparse);
+        pImpl->sqSolver->updateGradient(gradient_colMajor);
+    }
+
+
+    // Solve
+    if(!pImpl->sqSolver->isInitialized()) {
+        if(!pImpl->sqSolver->initSolver()) {
+            bfError << "OSQP: initSolver() failed.";
+            return false;
+        }
+    }
+
+    bool solveReturnVal = pImpl->sqSolver->solve();
+
+    if (pImpl->stopWhenFails && !solveReturnVal) {
+        bfError << "OSQP: solve() failed.";
+        return false;
+    }
+
+    // Get outputs
+    if (solveReturnVal) {
+        solution_colMajor = pImpl->sqSolver->getSolution();
+    }
```

Comment:
Maybe from here on we can assume that `solveReturnVal` is true, right? This if seems redundant.

Reply:
No, if `stopWhenFails` is false you don't know if `solveReturnVal` is true or false.

Reply:
Ok now I got the logic, makes sense!